File: CombinationSum-Leetcode.java
#java
class Solution {
    
    public void printcombination(int i,int[] arr, int t, List<List<Integer>> ans , List<Integer>ds){
        //base case
        if(i==arr.length){
            if(t==0){
                ans.add(new ArrayList<>(ds));
            }
            return;
        }        
        //pick
        if(arr[i]<=t){
        ds.add(arr[i]);
         printcombination(i,arr,t-arr[i],ans,ds);
        ds.remove(ds.size() -1);
        }
        //not pick
        printcombination(i+1,arr,t,ans,ds);
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans=new ArrayList<>();
        printcombination(0,candidates,target,ans,new ArrayList<Integer>());
        return ans;
    }
}

File: CombinationSum-Leetcode.cpp
#c++
class Solution {
public:
    
    void printcombi(int i,vector<int>& arr,int t,vector<vector<int>>& ans , vector<int>& ds ){
        if(i==arr.size()){
            if(t==0)
                ans.push_back(ds);
            return ;
        }
        
        //pick
        if(arr[i]<=t){
            ds.push_back(arr[i]);
            printcombi(i,arr,t-arr[i],ans,ds);
            ds.pop_back();
        }
        //dont pick
        printcombi(i+1,arr,t,ans,ds);
        
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> ds;
        printcombi(0,candidates,target,ans,ds);
        return ans;
    }
};
